bigquery-antipattern-recognition/src/main/java/com/google/zetasql/toolkit/antipattern/rewriter/gemini/QueryVisitorRewriter.java

```diff
@@ -0,0 +1,113 @@
+package com.google.zetasql.toolkit.antipattern.rewriter.gemini;
+
+import com.google.zetasql.SqlException;
+import com.google.zetasql.toolkit.antipattern.AntiPatternVisitor;
+import com.google.zetasql.toolkit.antipattern.analyzer.visitors.joinorder.JoinOrderVisitor;
+import com.google.zetasql.toolkit.antipattern.cmd.InputQuery;
+import com.google.zetasql.toolkit.antipattern.exceptions.TTLExpiredDuringRewriteException;
+import com.google.zetasql.toolkit.antipattern.rewriter.prompt.PromptYamlReader;
+import com.google.zetasql.toolkit.antipattern.util.AntiPatternHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class QueryVisitorRewriter {
+    private static final Logger logger = LoggerFactory.getLogger(QueryVisitorRewriter.class);
+    private final PromptYamlReader promptYamlReader;
+    private final AntiPatternHelper antiPatternHelper;
+    private final Boolean llmStrictValidation;
+
+    public QueryVisitorRewriter(AntiPatternHelper antiPatternHelper, boolean llmStrictValidation) throws IOException {
+        this.promptYamlReader = new PromptYamlReader();
+        this.antiPatternHelper = antiPatternHelper;
+        this.llmStrictValidation = llmStrictValidation;
+
+    }
+
+    // Retries will be used:
+    // 1- In case the antipattern is still there
+    // 2- In case the query is syntactically invalid
+    public String rewriteSQL(String inputQuery,
+                             AntiPatternVisitor visitorThatFoundAntiPattern,
+                             Integer llmRetries) throws Exception {
+        String prompt = this.promptYamlReader.getAntiPatternNameToPrompt().get(visitorThatFoundAntiPattern.getName());
+
+        if (prompt == null) {
+            return  inputQuery; // No changes so it can use the same query
+        }
+
+        prompt = String.format(prompt, inputQuery).replace("%%","%");
+        String queryStr = GeminiRewriter.processPrompt(prompt, this.antiPatternHelper.getProject());
+        queryStr = checkAntiPattern(queryStr, visitorThatFoundAntiPattern, llmRetries);
```

Comment:
we should only run this check if llmRetries > 0 and llmStrictValidation != null.

Currently we are validation regardless of the values of those vars and then raising an exception if validation failed and llmStrictValidation!=0 and llmRetries > 0. This leads to potentially 1 extra validation per each query per each visitor that found anti pattern. And it also sets a baseline of at least 1 validation per each query per each visitor that found anti pattern.

Reply:
all i am saying is that we should apply the filters check `llmRetries > 0 and this.llmStrictValidation is not null` at the beginning of checkAntiPattern. That way we save ourselves and additional visit per query per visitor that found anitpattern

Reply:
I do understand I just don't agree, it is not a unreasonable request though, I'll make the change.

Reply:
Done